src/cmd/services/m3coordinator/ingest/ingest.go

```diff
@@ -0,0 +1,196 @@
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package ingest
+
+import (
+	"context"
+	"time"
+
+	"github.com/m3db/m3/src/cmd/services/m3coordinator/server/m3msg"
+	"github.com/m3db/m3/src/query/models"
+	"github.com/m3db/m3/src/query/storage"
+	"github.com/m3db/m3/src/query/ts"
+	"github.com/m3db/m3/src/x/serialize"
+	"github.com/m3db/m3metrics/metric/id"
+	"github.com/m3db/m3metrics/policy"
+	xerrors "github.com/m3db/m3x/errors"
+	"github.com/m3db/m3x/instrument"
+	"github.com/m3db/m3x/pool"
+	"github.com/m3db/m3x/retry"
+	xsync "github.com/m3db/m3x/sync"
+	xtime "github.com/m3db/m3x/time"
+
+	"github.com/uber-go/tally"
+)
+
+// Options configures the ingester.
+type Options struct {
+	Appender          storage.Appender
+	Workers           xsync.PooledWorkerPool
+	PoolOptions       pool.ObjectPoolOptions
+	TagDecoderPool    serialize.TagDecoderPool
+	RetryOptions      retry.Options
+	InstrumentOptions instrument.Options
+}
+
+type ingestMetrics struct {
+	ingestError   tally.Counter
+	ingestSuccess tally.Counter
+}
+
+func newIngestMetrics(scope tally.Scope) ingestMetrics {
+	return ingestMetrics{
+		ingestError:   scope.Counter("ingest-error"),
+		ingestSuccess: scope.Counter("ingest-success"),
+	}
+}
+
+// Ingester ingests metrics with a worker pool.
+type Ingester struct {
+	workers xsync.PooledWorkerPool
+	p       pool.ObjectPool
+}
+
+// NewIngester creates an ingester.
+func NewIngester(
+	opts Options,
+) *Ingester {
+	retrier := retry.NewRetrier(opts.RetryOptions)
+	m := newIngestMetrics(opts.InstrumentOptions.MetricsScope())
+	p := pool.NewObjectPool(opts.PoolOptions)
+	p.Init(
+		func() interface{} {
+			// NB: we don't need a pool for the tag decoder since the ops are
+			// pooled, but currently this is the only way to get tag decoder.
+			tagDecoder := opts.TagDecoderPool.Get()
+			op := ingestOp{
+				s:  opts.Appender,
+				r:  retrier,
+				it: serialize.NewMetricTagsIterator(tagDecoder, nil),
+				p:  p,
+				m:  m,
+				c:  context.TODO(),
+			}
+			op.attemptFn = op.attempt
+			op.ingestFn = op.ingest
+			return &op
+		},
+	)
+	return &Ingester{
+		workers: opts.Workers,
+		p:       p,
+	}
+}
+
+// Ingest ingests a metric asynchronously with callback.
+func (i *Ingester) Ingest(
+	id []byte,
+	metricTime time.Time,
+	value float64,
+	sp policy.StoragePolicy,
+	callback *m3msg.RefCountedCallback,
```

Comment:
Can we add `ctx context.Context` as the first argument?  (lint forces some context to always be the first arg)

We don't want to use the TODO context each time.

Reply:
hmm, the m3msg server will only be able to pass a `TODO`/`Background` context each time though, although I can reuse one context there, is it what you wanted?

Reply:
Yes I think that's ideal, in the future if we have a per-op context then we'll be able to use that.  Otherwise here it'll be a single one that can never be specialized.  We should at least let the callers into this layer be able to specify their own.